### PR TITLE
Inform serial protocols of missing packets

### PR DESCRIPTION
**Pull request by pkendall64**
# Problem
When using SBUS at lower packet rates, 50Hz for example, each second packet (approximately) has the "lost frame" bit set because the SBUS packets are sent at a regular cadence of 9ms. So each second SBUS output has not received a new packet over-the-air.

# Solution
We know when we've missed a packet in the main RX code (we use it to calculate LQ), so also inform serial protocols that a packet has been missed. In this case, when the protocol is called and there is no new packet and no missed packet then it can just ignore it's output window and wait till data is ready.
In the case of SBUS, it can use the missed frame flag to send out the packet and if no new data or missed frame is detected then it can delay outputting any data. So, at 50Hz SBUS will output packets at 20ms intervals. At higher rates (i.e. 150+ Hz) it will output at the regular 9ms interval for each output packet, effectively ignoring some over-the-air packets.

### CODE REVIEW

File: src/src/rx-serial/SerialSBUS.cpp [hunk 1]
```diff
@@ -8,19 +8,23 @@
...
 const auto SBUS_CALLBACK_INTERVAL_MS = 9;
 
-uint32_t SerialSBUS::sendRCFrame(bool frameAvailable, uint32_t *channelData)
+uint32_t SerialSBUS::sendRCFrame(bool frameAvailable, bool frameMissed, uint32_t *channelData)
 {
     static auto sendPackets = false;
     if ((failsafe && config.GetFailsafeMode() == FAILSAFE_NO_PULSES) || (!sendPackets && connectionState != connected))
     {
-        return SBUS_CALLBACK_INTERVAL_MS;
+        return UNCONNECTED_CALLBACK_INTERVAL_MS;
```
Comment (JyeSmith) [inline src/src/rx-serial/SerialSBUS.cpp:19]: Why has this changed?

File: src/src/rx-serial/SerialSBUS.cpp [hunk 2]
```diff
@@ -8,19 +8,23 @@
...
     }
     sendPackets = true;
 
+    if ((!frameAvailable && !frameMissed) || _outputPort->availableForWrite() < 25)
```
Comment (JyeSmith) [inline src/src/rx-serial/SerialSBUS.cpp:23]: `availableForWrite` is a nice check.  Did you see issues with the buffer filling?

File: src/src/rx-serial/SerialSBUS.cpp (hunk 1 above)
Comment (pkendall64) [reply src/src/rx-serial/SerialSBUS.cpp:19]: Just using a more appropriately named constant.

File: src/src/rx-serial/SerialSBUS.cpp (hunk 2 above)
Comment (pkendall64) [reply src/src/rx-serial/SerialSBUS.cpp:23]: No issues, just being safe. I don't want it to block and also don't want to be pushing partial packets.

**Review by JyeSmith — APPROVED** [review]

**Review by CapnBry — APPROVED** [review]
I had to draw out a little timing diagram in my notebook to make sure I understood that frames or the missed flag would not be lost in any condition / packet rate but this all checks out.

I'm not super pleased about 50Hz running balls-to-the-wall for 11ms between packets but at 50Hz I guess the MCU has a lot of spare time on its hands anyway. It is an elegant solution to sync the output to 50Hz though so I'm on board.

**Pull request merged**